rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,258 @@
+    %add = stablehlo.add %arg1, %arg2 : tensor<i32>
+    stablehlo.return %add : tensor<i32>
+  }) {dimensions = dense<[0]> : tensor<1xi64>}
+    : (tensor<?xi32>, tensor<i32>) -> tensor<i32>
+  func.return %sum : tensor<i32>
+}
+```
+
+Originally, HLO introduced `set_dimension_size` op as it neither had dynamic
+types nor dynamic ops. StableHLO dialect doesn't have these limitations and
+therefore new users don't need to make use of this low level op unless they are
+moving from HLO or MHLO to StableHLO. The TensorFlow and JAX teams believe that
+this hypothesis should be correct based on their experiences so far. Dynamic
+operations `real_dynamic_slice` and `dynamic_pad` can be used instead.
+
+The following example demonstrates the differences between programs using
+unbounded dynamism and bounded dynamism.
+
+```mlir
+func.func @slice_with_unbounded_dynamism(%data: tensor<7xf32>, %start: tensor<1xi32>, %limit: tensor<1xi32>) -> tensor<?xf32> {
+  %strides = stablehlo.constant dense<1> : tensor<1xi32>
+  %result =  stablehlo.real_dynamic_slice %data, %start, %limit, %strides
+    : (tensor<7xf32>, tensor<1xi32>, tensor<1xi32>, tensor<1xi32>)
+      -> tensor<?xf32>
+  func.return %result : tensor<?xf32>
+}
+```
+
+```mlir
+func.func @slice_with_bounded_dynamism(%data: tensor<7xf32>, %start: tensor<1xi32>, %limit: tensor<1xi32>) -> tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>> {
+  // Add padding to avoid OOM access in the following slice op.
+  %pad_value = stablehlo.constant dense<0.0> : tensor<f32>
+  %padded_data = stablehlo.pad %data, %pad_value, low = [0], high = [7], interior = [0]
+    : (tensor<7xf32>, tensor<f32>) -> tensor<14xf32>
+
+  // Extract the largest possible slice starting at the start index.
+  %scalar_start = stablehlo.reshape %start : (tensor<1xi32>) -> tensor<i32>
+  %padded_result = stablehlo.dynamic_slice %padded_data, %scalar_start, sizes = [7]
+    : (tensor<14xf32>, tensor<i32>) -> tensor<7xf32>
+
+  // Remove the extra elements extracted beyond the limit.
+  %slice_size = stablehlo.subtract %limit, %start : tensor<1xi32>
+  %scalar_size = stablehlo.reshape %slice_size : (tensor<1xi32>) -> tensor<i32>
+  %result = stablehlo.set_dimension_size %padded_result, %scalar_size, dim = 0
+    : (tensor<7xf32>, tensor<i32>)
+      -> tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>>
+
+  func.return %result : tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>>
+}
+
+```
+
+Use of dynamic ops over `set_dimension_size` op has various benefits:
+
+* Greatly simplifies the lowering from higher level frameworks to StableHLO as
+  they don't need to compute upper bounds of dynamic dimensions.
+* Makes lowerings to StableHLO hardware agnostic as they don't depend on if the
+  hardware requires unbounded or bounded programs.
+* Reduces the potential confusion in making use of `set_dimension_size` as the
+  users are generally not familiar with this op and also the semantics are not
```

Comment:
This comment is debatable. Suggesting to include advantages that are mostly matter-of-fact than matter-of-interpretation.

Reply:
Removed. If there is someone disagreeing with it, it might not be broadly true. :)